### ISSUE TITLE
Add constraints from julia process into Conda version resolution (somehow?)

### ISSUE THREAD

**Issue by ericphanson** [cjdoris/CondaPkg.jl#43]
Context: it turns out scipy v1.9 adds a C++ dependency on HiGHs that seems incompatible with Julia 1.7 on ubuntu (works on my m1 macbook). E.g. a CondaPkg toml with
```
[deps]
scikit-learn = ">=1.1.0,<1.2"
scipy = ">=1.0,<1.9"
```
will let `pyimport("sklearn")` succeed, while just
```
[deps]
scikit-learn = ">=1.1.0,<1.2"
```
will fail, assuming it gets scipy v1.9 on ubuntu:
```julia
julia> sklearn=pyimport("sklearn")
ERROR: Python: ImportError: /home/ubuntu/.asdf/installs/julia/1.7.3/bin/../lib/julia/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /home/ubuntu/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/optimize/_highs/_highs_wrapper.cpython-310-x86_64-linux-gnu.so)
Python stacktrace:
 [1] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/optimize/_linprog_highs.py:20
 [2] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/optimize/_linprog.py:21
 [3] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/optimize/__init__.py:413
 [4] import_module
   @ importlib ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/importlib/__init__.py:126
 [5] __getattr__
   @ scipy ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/__init__.py:211
 [6] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:24
 [7] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/stats/distributions.py:8
 [8] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/stats/_stats_py.py:46
 [9] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/scipy/stats/__init__.py:467
 [10] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/sklearn/utils/fixes.py:19
 [11] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/sklearn/utils/__init__.py:29
 [12] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/sklearn/base.py:17
 [13] <module>
   @ ~/projects/LighthouseWandb.jl/.CondaPkg/env/lib/python3.10/site-packages/sklearn/__init__.py:82
```

My very limited understanding of things is that Julia itself imposes some compatibility constraints on what python packages can be loaded into the process, and that these constraints aren't known to Conda at version resolution time, and therefore we can get incompatibilites. I wonder if there's a way to tell Conda we need a certain GCC version, or to have a dummy Julia package in the Conda env that somehow can express this compat.

**Comment by cjdoris** [comment]
AFAIU Julia already has some libraries linked when it starts up, and these can shadow ones from Python packages, hence the incompatibility.

There's probably something we can do, but the hard part will be figuring out what the constraints should even be - I imagine they can change across Julia versions and operating systems.

Does the incompatibility really lie with SciPy or some dependency? What is the diff between the packages that are installed in the two configurations above?

**Comment by ericphanson** [comment]
I am pretty sure it is scipy because v1.9 added HiGHs for the first time (https://docs.scipy.org/doc/scipy/release.1.9.0.html#scipy-optimize-improvements) which is a C++ optimization solver that shows up in the error message above.

**Comment by cjdoris** [comment]
OK I've done some digging and thinking.

The constraint really is on libstdc++: SciPy 1.9 requires libstdc++ 12 but Julia 1.7 is built with an earlier libstdc++ (v11 I believe, looking at https://gcc.gnu.org/onlinedocs/gcc-12.1.0/libstdc++/manual/manual/abi.html).

So one solution is to have the Conda dependency `libstdcxx-ng<12` in PythonCall (you can verify yourself that this will restrict to SciPy 1.8). But some issues with this are:
- The actual constraint depends on the version of Julia and the platform. How do we work this out reliably (without it becoming a burden to maintain)? Can we get the libstdc++ version Julia was build with from within Julia? Or is there an external source of this information?
- It will install a package (libstdcxx-ng) that is unnecessary for some users. Not a big problem.
- CondaPkg dependencies are declared statically, so cannot depend on the platform. One solution is to add a build script to PythonCall which adds the dependency. Or we just have a single dependency covering all supported versions of Julia.

**Comment by ericphanson** [comment]
> Can we get the libstdc++ version Julia was build with from within Julia? Or is there an external source of this information?

I asked about this a bit on the `#binarybuilder` channel on the Julia Slack. There @keno suggested inspecting the version of the CompilerSupportLibraries_jll stdlib from within Julia and having a handwritten lookup table of CompilerSupportLibraries_jll versions to GCC versions.

I think the way to get the version from within Julia would be
```julia
julia> using TOML, CompilerSupportLibraries_jll

julia> TOML.parsefile(joinpath(pkgdir(CompilerSupportLibraries_jll), "Project.toml"))["version"]
"0.5.2+0"
```

I am not really sure how to build the lookup table though. CompilerSupportLibraries_jll itself is built here: https://github.com/JuliaPackaging/Yggdrasil/blob/master/C/CompilerSupportLibraries/build_tarballs.jl. And that script has the version number of CompilerSupportLibraries_jll that it is building. But the GCC version isn't similarly hardcoded, so my best guess is one needs to look at when the version number was bumped and try to figure out what GCC version was being used at that time. That seems very manual and error prone though.

> CondaPkg dependencies are declared statically, so cannot depend on the platform. One solution is to add a build script to PythonCall which adds the dependency. Or we just have a single dependency covering all supported versions of Julia.

This seems tricky. I guess one starting point could just be clear documentation explaining the issue and suggesting adding a `libstdcxx-ng` compat bound (ideally with some information on how to figure out what the bound should be), without doing anything automatic.

**Comment by ericphanson** [comment]
I think something like this would be really useful; it took me awhile to figure out why something was broken, and the error doesn't always surface in the nicest way.

E.g. for me it came up as the wandb python library depends on sklearn if you want to log ROC curves (so it's kinda an optional dep), and the error surfaced to me was "Install sklearn if you want to use ROC curves". Well I thought I was installing sklearn and it took some debugging to reproduce (e.g. only showed up on linux, not my mac laptop) and dig in to figure out actually sklearn was being installed but it was erroring at runtime because of it's dependency on scipy and this clash, and the wandb was swallowing the error.

So in other words, this kind of error is platform dependent so it can be hard to repro, and can show up unexpectedly, so I think any safety CondaPkg can add around this will be really helpful.

In terms of the implementation, I think it could be nice if it was configurable somehow, because I could see folks w custom Julia builds being annoyed, but I imagine 99% of users are using a stock julia build and adding the compat is good. So anyway maybe it's ok to assume stock for now and see if anyone complains 😄.

**Comment by cjdoris** [comment]
Thanks that's useful feedback. I'll continue to mull it over.

As for making this configurable, I bet the intersection between people using custom Julia builds, people using PythonCall, and people needing a more recent libstdc++ than this allows is quite small! So probably not pressing. I've been wanting to add a generic way for the user to over-ride CondaPkg dependencies somehow, which would solve this generically without adding more config to PythonCall.

Tangentially, how well does W&B work from PythonCall? I assume basic metric logging works OK, but what about other "magic" stuff like logging the command, the script, the git commit, process-level stuff, etc.?

**Comment by ericphanson** [comment]
> Tangentially, how well does W&B work from PythonCall? I assume basic metric logging works OK, but what about other "magic" stuff like logging the command, the script, the git commit, process-level stuff, etc.?

I mostly just use the metric logging, basic metrics but also plots and such. The command gets logged out as `<python with no main file>` but it does have the command-line arguments for the julia process. The git seems to work fine. It also tracks stuff like CPU usage, memory, gpu usage etc, no problem. I haven't tried the source code tracking stuff. We already had a pretty robust system for reproducibility/versioning/etc so for me the main purpose is logging and plots and reports to compare training runs, and for that it's been working fine. I would say for me the biggest annoyance of W&B-in-julia is the python dependency itself (which PythonCall/CondaPkg help a lot with!) just due to install issues like this, and slowing down docker builds etc.

**Comment by cjdoris** [comment]
I have just released a version of PythonCall with this fix in, so closing the issue.

It's cool that W&B works so well from Julia! I'm quite amazed actually.
